src/server/modules/threat/services/SafeBrowsingService.ts

```diff
@@ -89,6 +90,76 @@ export class SafeBrowsingService implements UrlThreatScanService {
     }
     return matches
   }
+
+  // TODO: deduplicate this function with the lookup function above
+  // ideally, we should do a non-block write for any matched viruses
+  // into the redis cache. the write above is blocking so we chose to
+  // bypass the write for now
+  private async lookupBulk(urls: string[]) {
+    let matches = null
+    const request = { ...this.requestTemplate } as any
+
+    request.threatInfo.threatEntries = urls.map((url) => {
+      return { url }
+    })
+
+    const response = await fetch(ENDPOINT, {
+      method: 'POST',
+      headers: { 'Content-Type': 'application/json' },
+      body: JSON.stringify(request),
+    })
+    if (!response.ok) {
+      const error = new Error(
+        `Safe Browsing failure:\tError: ${response.statusText}\thttpResponse: ${response}\t body:${response.body}`,
+      )
+      if (safeBrowsingLogOnly) {
+        logger.error(error)
+      } else {
+        throw error
+      }
+    } else {
+      const result = await response.json()
+      if (result?.matches) {
+        const prefix = safeBrowsingLogOnly
+          ? 'Considered threat by Safe Browsing but ignoring'
+          : 'Malicious link content'
+        result.matches.forEach((threatMatch: any) => {
+          logger.warn(
+            `${prefix}: ${threatMatch.threat.url} yields ${JSON.stringify(
+              threatMatch,
+              null,
+              2,
+            )}`,
+          )
+        })
+        matches = result.matches
+      }
+    }
+    return matches
+  }
+
+  public isThreatBulk: (urls: string[]) => Promise<boolean> = async (urls) => {
+    const urlChunks: string[][] = []
+
+    for (let i = 0; i < urls.length; i += SAFE_BROWSING_LIMIT) {
+      urlChunks.push(urls.slice(i, i + SAFE_BROWSING_LIMIT))
+    }
+
+    if (!safeBrowsingKey) {
+      logger.warn(`No Safe Browsing API key provided. Not scanning in bulk`)
+      return false
+    }
+
+    for (let i = 0; i < urlChunks.length; i += 1) {
```

Comment:
nit: if we are looking at more than 1000 links, Promise.all can potentially be faster. https://stackoverflow.com/questions/45285129/any-difference-between-await-promise-all-and-multiple-await
But highly optional IMO

Reply:
Hm..we chose to do this so that if any of them was malicious, it would terminate early instead of waiting for all the promises to complete. But since Promise.all will probably run concurrently.. I guess the difference is small. 

I can add it inline as a comment to look further into!

Reply:
[5b76d97](https://github.com/opengovsg/GoGovSG/pull/1993/commits/5b76d972f2250e6b5a019d858632b916d9e687a4)